### PR TITLE
Support configurable index urls for pip and private registry auth (not just) for pip

### PR DESCRIPTION
**Pull request by chmeliik**
TLDR:

* `--index-url` in requirements.txt is no longer ignored
* If the index url is anything other than https://pypi.org/simple/, report it in the SBOM
* Tell aiohttp not to ignore `.netrc`
* Add way to run local PyPI server that requires auth, add integration test that uses it

Run new test with:

```shell
make test-integration TEST_LOCAL_PYPISERVER=true TOX_ARGS='-k custom_index'
```

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

**Note:** if the contribution is external (not from an organization member), the CI
pipeline will not run automatically. After verifying that the CI is safe to run:

- [approve GitHub Actions workflows][approve-gh-actions] by clicking a button
- approve the Red Hat Trusted App Pipeline container build by commenting `/ok-to-test`
  (as is the [standard for Pipelines as Code][pac-running-pipeline])

[approve-gh-actions]: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
[pac-running-pipeline]: https://pipelinesascode.com/docs/guide/running/#running-the-pipelinerun

### CODE REVIEW

**Comment by chmeliik** [comment]
This should be everything that's needed for cachi2 to support custom PyPI servers, including auth.

Auth is done via .netrc, same as pip: https://pip.pypa.io/en/stable/topics/authentication/#netrc-support.
Pip also supports keyring auth but the docs make it sound rather strange. Doesn't seem worth supporting.

Let me know if you would like an ADR for any of this :saluting_face: 

TODO: make the integration test run in CI, ideally without copy-pasting most of the Makefile target

**Comment by taylormadore** [comment]
Wow, this did indeed end up being quite straightforward. LGTM so far :+1:

**Comment by chmeliik** [comment]
I've made the CI work with the new int. test, should be ready for review

File: Makefile [hunk 1]
```diff
@@ -32,7 +33,14 @@ test-unit: venv
 	venv/bin/tox -e $(TOX_ENVLIST) -- $(TOX_ARGS)
 
 test-integration: venv
-	CACHI2_GENERATE_TEST_DATA=$(GENERATE_TEST_DATA) venv/bin/tox -e integration
+	if [ $(TEST_LOCAL_PYPISERVER) = true ]; then \
+		PYPISERVER_PORT=$(PYPISERVER_PORT) PYPISERVER_IMAGE=$(PYPISERVER_IMAGE) tests/pypiserver/start.sh & \
+		trap "kill $$!" EXIT && \
+		while true; do sleep 1; curl -fsS http://localhost:$(PYPISERVER_PORT) && break; done; \
+	fi && \
```
Comment (eskultety) [inline Makefile:40]: I am very much not a fan of abusing the Makefile for this purpose (particularly to use Bash as a crutch instead of modeling this properly in the Python test suite). It really only should be a recipe for tedious build/test/release processes. Now that said, one might argue that this indeed is related to the "test" part I mentioned in the previous sentence, but not quite - configuring tox is a good example of that. If we optionally need a test server deployed for a particular test's needs then I believe we should do it part of the test suite and not part of the Makefile.
Related to your `integration/conftest.py` changes, is there anything preventing us to model this server as a test fixture with a cleanup section based on Python's subprocess? After all, you guard the test with a pytest  `skip` mark, so such a fixture should not run in that case and it would be nicely contained within the test suite where it belongs.

Bonus points if we could reuse the `utils.ContainerImage` class to do this which I think it should be possible looking at the command line the container gets.

File: tests/integration/conftest.py [hunk 2]
```diff
@@ -16,19 +16,20 @@ def test_data_dir() -> Path:
...
-        cachi2 = utils.build_image(cachi2_repo_root, tag="localhost/cachi2:latest")
-    else:
-        cachi2 = utils.ContainerImage(cachi2_image_ref)
-        if not cachi2_image_ref.startswith("localhost/"):
-            cachi2.pull_image()
+        utils.build_image(cachi2_repo_root, tag=cachi2_image_ref)
+
+    cachi2 = utils.Cachi2Image(cachi2_image_ref)
+    if not cachi2_image_ref.startswith("localhost/"):
+        cachi2.pull_image()
```
Comment (eskultety) [inline tests/integration/conftest.py:33]: nitpick: This whole hunk seems a little bit out of place in this commit, there's already quite a bit happening and this is easily missed, so I think this adjustment along with the `utils.ContainerImage` (not `utils.Cachi2Image` though) deserve their own spotlight in a separate commit.

File: tox.ini [hunk 3]
```diff
@@ -62,6 +62,9 @@ commands =
 passenv =
     CACHI2_GENERATE_TEST_DATA
     CACHI2_IMAGE
+    CACHI2_TEST_LOCAL_PYPISERVER
+setenv =
+    CACHI2_TEST_NETRC_CONTENT={env:CACHI2_TEST_NETRC_CONTENT:machine localhost login cachi2-user password cachi2-pass}
```
Comment (eskultety) [inline tox.ini:67]: Why do we want to make `CACHI2_TEST_NETRC_CONTENT` variable configurable from the outside for test purposes? Isn't a fixed value enough? When would we want to reconfigure the test differently?

**Review by eskultety — COMMENTED** [review]
Conceptually, no objections. I left a few suggestions and nitpick comments.

File: cachi2/core/package_managers/pip.py [hunk 4]
```diff
@@ -1853,7 +1860,9 @@ def _process_package_distributions(
     :rtype: list[DistributionPackageInfo]
     """
     allowed_distros = ["sdist", "wheel"] if allow_binary else ["sdist"]
-    client = pypi_simple.PyPISimple()
+    client = pypi_simple.PyPISimple(
+        endpoint=index_url if index_url else pypi_simple.PYPI_SIMPLE_ENDPOINT
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachi2/core/package_managers/pip.py]: Good call, changed

File: Makefile (hunk 1 above)
Comment (chmeliik) [reply Makefile:40]: I can try that, but I'd like to keep the `pypiserver-run` makefile target as well

The process will need to run in the background, not sure how pretty that will be

Comment (eskultety) [reply Makefile:40]: > The process will need to run in the background, not sure how pretty that will be

I'm not sure I follow, `subprocess.Popen` doesn't block the main process, IIUC if you put the Popen call to a fixture followed by a `yield` and then for the "cleanup" kill and wait for the server process that should theoretically work, shouldn't it?

Comment (chmeliik) [reply Makefile:40]: > pytest eats the stdout and stderr from the subprocess

Although with `--capture=sys` or `--capture=tee-sys` passed to pytest, this looks better. Still need to find out what those do though

Comment (eskultety) [reply Makefile:40]: Sigh, true. I guess if only

>  if the server fails to start, you'll never find out why (the output is logged when the tests finish

needs to be solved, I think redirecting podman's stdout to a file instead of the parent's process stdout (which pytest consumes) then that should also work, but then

> but it's already a bit worse than bash

will still remain true I suppose. Can we at least extract the Makefile bits, stuff it in a bash script and bury it somewhere under the integration test directory hierarchy? That shouldn't have an effect on the github action you needed the dedicated makefile target for.

Comment (eskultety) [reply Makefile:40]: > Although with --capture=sys or --capture=tee-sys passed to pytest, this looks better. Still need to find out what those do though

Looking at the docs `tee-sys` reads exactly like what we need.

Comment (chmeliik) [reply Makefile]: Added a WIP commit for what it would look like, see if you like it better

File: tests/integration/test_pip.py [hunk 5]
```diff
@@ -1,15 +1,48 @@
...
+        except requests.RequestException as e:
+            log.debug(e)
+    else:
+        proc.terminate()
+        raise RuntimeError("pypiserver didn't start fast enough")
+
+    yield
+    proc.terminate()
+
+
```
Comment (eskultety) [inline tests/integration/test_pip.py:45]: nitpick: not a big deal but strictly from correctness POV we should clean up the zombie process explicitly despite either Python runtime or whatever the parent is will take care of it.

File: Makefile (hunk 1 above)
Comment (eskultety) [reply Makefile]: > Added a WIP commit for what it would look like, see if you like it better

Unsurprisingly, it led to more lines in Python, but the code is still perfectly readable and contained within the test suite, nicely done!

File: tests/integration/test_pip.py (hunk 5 above)
Comment (chmeliik) [reply tests/integration/test_pip.py]: fixed

File: tests/integration/conftest.py (hunk 2 above)
Comment (chmeliik) [reply tests/integration/conftest.py:37]: Ack, split the changes

File: tox.ini (hunk 3 above)
Comment (chmeliik) [reply tox.ini:69]: I used it for example to see what happens when the password is wrong. It's not strictly necessary, but I would expect the tests to respect the env var if I try to override it

**Comment by chmeliik** [comment]
One more small improvement: used contextlib.ExitStack to make absolutely sure the process get `.terminate()`-ed before python starts waiting for it to clean it up

File: tests/integration/conftest.py [hunk 6]
```diff
@@ -28,8 +33,39 @@ def cachi2_image() -> utils.ContainerImage:
...
+
+
+# autouse=True: It's nicer to see the pypiserver setup logs at the beginning of the test suite.
+# Otherwise, pypiserver would start once the pip tests need it and the logs would be buried between
+# test output.
+@pytest.fixture(autouse=True, scope="session")
+def local_pypiserver() -> Iterator[None]:
+    if os.getenv("CACHI2_TEST_LOCAL_PYPISERVER") != "true":
+        yield
+        return
```
Comment (eskultety) [inline tests/integration/conftest.py:50]: You can use `mark.skipif` on fixtures as well which will enhance readability by just that tiny bit here :).

File: Makefile [hunk 7]
```diff
@@ -32,6 +32,9 @@ test-unit: venv
 test-integration: venv
 	CACHI2_GENERATE_TEST_DATA=$(GENERATE_TEST_DATA) venv/bin/tox -e integration
 
+pypiserver-run:
+	tests/pypiserver/start.sh
+
```
Comment (eskultety) [inline Makefile]: I'm still not convinced by this target, even more so if with all the latest changes this was reduced to a mere invocation of a script. I'd encourage it if the pypi server were to serve as a development deployment of a crucial component to the project, but it's not, we only need it for a specific test which I expect will be the only use case for a long time. If anything, we could document this as part of the integration test suite section mentioning that one can use the script to run a server locally.

File: tests/pypiserver/start.sh [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+WORKDIR=$(mktemp -d --tmpdir "cachi2-pypiserver.XXXXXX")
+trap cleanup EXIT
+
+setup() {
+    echo -e "\n--- Downloading $DIR/package-urls.txt ---\n"
+
+    mkdir "$WORKDIR/packages"
+    sed '/^#/d' "$DIR/package-urls.txt" |
```
Comment (eskultety) [inline tests/pypiserver/start.sh:22]: wouldn't `tail -n +2` to skip the first line be actually more straightforward and slightly more readable?

File: tests/integration/conftest.py [hunk 9]
```diff
@@ -19,16 +19,17 @@ def test_data_dir() -> Path:
 def cachi2_image() -> utils.ContainerImage:
     cachi2_image_ref = os.environ.get("CACHI2_IMAGE")
```
Comment (eskultety) [inline tests/integration/conftest.py:25]: **Commit message:**
`s/once/one`     perhaps?

**Review by eskultety — COMMENTED** [review]
Looking good, a couple more nitpics and I think we're ready to go.

**Comment by chmeliik** [comment]
Just a rebase, will address comments in next force-push

File: Makefile (hunk 7 above)
Comment (chmeliik) [reply Makefile]: Ack, removed and added to the README instead

File: tests/integration/conftest.py (hunk 6 above)
Comment (chmeliik) [reply tests/integration/conftest.py:50]: Tried this, but it doesn't seem to work https://github.com/pytest-dev/pytest/issues/964

Also tried `pytest.skip` instead of `yield; return`, but it looks like that also skips the tests that rely on this fixture (and together with `autouse=True`, that means skip all tests :smile: )

File: tests/pypiserver/start.sh (hunk 8 above)
Comment (chmeliik) [reply tests/pypiserver/start.sh:22]: I went with a "comment" syntax, in case I wanted to add more packages to download (and more comments explaning how I got the urls)

File: tests/integration/conftest.py (hunk 6 above)
Comment (eskultety) [reply tests/integration/conftest.py:50]: > Tried this, but it doesn't seem to work

sigh, I guess I forgot to remove the existing code when I tried it which falsely led me to thinking that it works, sorry.

**Review by eskultety — APPROVED** [review]

**Review by brunoapimentel — APPROVED** [review]
I did not test any external private registry, but the approach, code and tests LGTM.
